src/api/utils/request.js

```diff
@@ -0,0 +1,113 @@
+/* @flow */
+
+import type { $Request } from 'express';
+import FormData from 'form-data';
+import snakeCase from 'lodash/snakeCase';
+import fetch from 'node-fetch';
+
+import getResponseData from './getResponseData';
+import HttpError from './HttpError';
+import translateKeys from './translateKeys';
+
+type File = {
+  fieldname: string,
+  originalname: string,
+  buffer: Buffer,
+};
+
+export type Request = $Request & {
+  files?: File[],
+};
+
+export type Data = mixed & {
+  files?: string[],
+};
+
+export type RequestOptions = {
+  method: string,
+  url: string,
+  data?: ?Data,
+  request: Request,
+  authorization?: string,
+  getHeaders: Request => { [string]: string },
+};
+
+type Init = {
+  method: string,
+  headers: {
+    Authorization: ?string,
+    Accept: string,
+    'Content-Type'?: string,
+  },
+  body?: string | FormData,
+};
+
+export default async function request<T>({
+  method,
+  url,
+  data,
+  getHeaders,
+  authorization,
+  request: req,
+}: RequestOptions): Promise<?T> {
+  const init: Init = {
+    method,
+    headers: {
+      Authorization: authorization,
+      Accept: 'application/json',
+      ...getHeaders(request),
+    },
+  };
+
+  if (data) {
+    const reqFiles = req.files;
+
+    if (data.files && reqFiles) {
+      const { files, ...fields } = data;
+      const formData = new FormData();
+
+      Object.keys(fields).forEach(fieldName =>
+        formData.append(snakeCase(fieldName), fields[fieldName]),
+      );
+
+      files.forEach(fieldName => {
+        reqFiles
+          .filter(file => file.fieldname === fieldName)
+          .forEach(({ buffer, originalname }) =>
+            formData.append(snakeCase(fieldName), buffer, {
+              filename: originalname,
+            }),
+          );
+      });
+
+      init.body = formData;
+    } else {
+      init.headers['Content-Type'] = 'application/json';
+      init.body = JSON.stringify({
+        data: translateKeys(data, snakeCase),
+      });
+    }
+  }
+  const response = await fetch(url, init);
```

Comment:
this is another API trouble spot. BFLY clears out stuff for loggin here? seems like there is a better place for it but i'm not sure on the context.

In general i sort of hate this file, there is to much in here, that isn't sending an http request and it's deep deep down the module hierarchy for getting apis to it

Reply:
Maybe we can make the base method on `HttpApi` overrideable? Or require overriding it?